File: java_solutions/2303.calculate-amount-paid-in-taxes.java
/*
 * @lc app=leetcode id=2303 lang=java
 *
 * [2303] Calculate Amount Paid in Taxes
 */

class Solution {
    public double calculateTax(int[][] brackets, int income) {
        return IntStream.range(0, brackets.length)
            .mapToDouble(i -> Math.max(0, (Math.min(income, brackets[i][0]) - (i == 0 ? 0 : brackets[i - 1][0]))) * brackets[i][1] / 100.0)
            .sum();
    }
}

File: kotlin_solutions/2303.calculate-amount-paid-in-taxes.kt
/*
 * @lc app=leetcode id=2303 lang=kotlin
 *
 * [2303] Calculate Amount Paid in Taxes
 */

class Solution {fun calculateTax(brackets: Array<IntArray>, income: Int): Double = brackets.fold(Triple(0.0, income, 0)) { (ans, rem, lower), bracket -> if (rem <= 0) return ans else minOf(rem, bracket[0] - lower).let { amt -> Triple(ans + bracket[1] * amt / 100.0, rem - amt, bracket[0]) } }.first}
